### PR TITLE
1196 serialization

### PR DESCRIPTION
**Pull request by pmconrad**
Fixes part of https://github.com/bitshares/bitshares-core/issues/1196 .
This is a breaking that change requires additional changes in core. PR to follow.

Note that `boost::endian` is available from boost-1.58 on, so we will have to bump our minimum requirements.

Tested on an emulated big-endian PPC64 system (and x86 of course).

### CODE REVIEW

File: include/fc/io/raw.hpp [hunk 1]
```diff
@@ -89,52 +90,66 @@ namespace fc {
...
        uint64_t usec_as_int64;
-       s.read( (char*)&usec_as_int64, sizeof(usec_as_int64) );
+       unpack( s, usec_as_int64, _max_depth );
        usec = fc::microseconds(usec_as_int64);
     } FC_RETHROW_EXCEPTIONS( warn, "" ) }
 
     template<typename Stream, typename T, size_t N>
-    inline void pack( Stream& s, const fc::array<T,N>& v, uint32_t _max_depth ) {
-      s.write((const char*)&v.data[0],N*sizeof(T));
+    inline void pack( Stream& s, const fc::array<T,N>& v, uint32_t _max_depth ) = delete;
```
Comment (nathanielhourt) [inline include/fc/io/raw.hpp]: I'm getting a build error on this:

    ../libraries/fc/include/fc/io/raw.hpp:133:83: error: deleted definition must be first declaration
        inline void pack( Stream& s, const fc::array<T,N>& v, uint32_t _max_depth ) = delete;
                                                                                      ^
    ../libraries/fc/include/fc/io/raw_fwd.hpp:131:65: note: previous declaration is here
        template<typename Stream, typename T, size_t N> inline void pack( Stream& s, const fc::array<T,N>& v, uint32_t _max_depth=FC_PACK_MAX_DEPTH );
                                                                    ^


And similar on `unpack` below:

    ../libraries/fc/include/fc/io/raw.hpp:144:79: error: deleted definition must be first declaration
        inline void unpack( Stream& s, fc::array<T,N>& v, uint32_t _max_depth ) = delete;
                                                                                  ^
    ../libraries/fc/include/fc/io/raw_fwd.hpp:132:65: note: previous declaration is here
        template<typename Stream, typename T, size_t N> inline void unpack( Stream& s, fc::array<T,N>& v, uint32_t _max_depth=FC_PACK_MAX_DEPTH);
                                                                    ^

**Review by nathanielhourt — CHANGES_REQUESTED** [review]

**Comment by nathanielhourt** [comment]
Do we have any tests that were failing on big endian systems before, which are now working? If not, it would be nice to see some... perhaps serializing some hashes and public keys to strings and checking that they match expected values?

File: include/fc/io/raw.hpp (hunk 1 above)
Comment (nathanielhourt) [reply include/fc/io/raw.hpp]: Seems the solution is simply to delete the lines in `raw.hpp` and put the `= delete;` on the forward declarations in `raw_fwd.hpp`?

**Comment by pmconrad** [comment]
The first test that fails without modification is `fc_crypto/base36_test` because `fc::bigint(uint64_t)` is endian-dependent.

File: include/fc/io/raw.hpp (hunk 1 above)
Comment (pmconrad) [reply include/fc/io/raw.hpp]: Done.

**Comment by pmconrad** [comment]
Force-pushed after rebase to resolve conflicts

**Review by nathanielhourt — APPROVED** [review]
Looks fine to me.

**Pull request merged**